vllm fix check on max vocab size

xw285cornell opened the pull request:
Summary:
the tokenizer.vocab_size and model.vocab_size can be different. For QWen model, the tokenizer max token id is 151643 and the model config is `"vocab_size": 151936`. If we send an id between 151643 and 151936, it'll fail. Though in reality the tokenizer will just put ''. 

It's probably still valid to send the ids in between, because the model can legitimately produce such token id.

Test Plan:
Send 151860 and it's passing. Send 152860 and it complained about invalid token.

Rollback Plan:

Differential Revision: D79840114

22quinn commented on vllm/v1/engine/processor.py:
```diff
@@ -394,7 +394,7 @@ def _validate_model_input(
             tokenizer = self.tokenizer.get_lora_tokenizer(lora_request)
             max_input_id = max(prompt_ids, default=0)
-            if max_input_id > tokenizer.max_token_id:
+            if max_input_id > max(tokenizer.max_token_id, self.model_config.get_vocab_size()):
```
perhaps no need max? model config vocab size is the source of truth.
```suggestion
            if max_input_id > self.model_config.get_vocab_size() - 1:
```

houseroad replied:
Do we observe that we have such ids generated in Qwen?

xw285cornell replied:
Yes, we have such ID generated by Qwen which is out of the tokenizer's vocabulary.

houseroad replied:
What does that mean then? Just curious how the model / engine handle it?

22quinn replied:
Some explanation here: https://github.com/vllm-project/vllm/issues/13175#issuecomment-2781842571
and here why vocab size is 151936(%128==0): https://github.com/QwenLM/Qwen3/issues/147#issuecomment-1988398254

ywang96 replied:
@22quinn This suggestion wouldn't work with some multimodal models that use a token in the vocabulary of the tokenizer as placeholder token but such token is included from the model vocabulary (to prevent model to generate such placeholder token itself), so I think we still need `max` here since this is for input processing.

22quinn replied:
so in that case `tokenizer.max_token_id > self.model_config.get_vocab_size()`? Do you have an example?

xw285cornell commented:
@22quinn  @houseroad  updated as requested, do you want to take another look?

22quinn reviewed (approved):
LGTM

ywang96 reviewed (approved):
LGTM - FYI @WoosukKwon

houseroad commented on vllm/v1/engine/processor.py:
```diff
@@ -394,7 +394,7 @@ def _validate_model_input(
             tokenizer = self.tokenizer.get_lora_tokenizer(lora_request)
             max_input_id = max(prompt_ids, default=0)
-            if max_input_id > tokenizer.max_token_id:
+            if max_input_id > self.model_config.get_vocab_size() - 1:
```
maybe add some comments including the pointers from Quinn?

houseroad reviewed (approved):
Looks good.

DarkLight1337 commented:
Can you merge from main to fix the CI failures?

xw285cornell committed:
vllm fix check on max vocab size (#22471)

Summary:

the tokenizer.vocab_size and model.vocab_size can be different. For QWen model, the tokenizer max token id is 151643 and the model config is `"vocab_size": 151936`. If we send an id between 151643 and 151936, it'll fail. Though in reality the tokenizer will just put ''. 

It's probably still valid to send the ids in between, because the model can legitimately produce such token id.

Test Plan:
Send 151860 and it's passing. Send 152860 and it complained about invalid token.

Rollback Plan:

Reviewed By: tensormeta, houseroad

Differential Revision: D79840114

hijkzzz commented:
a serious bug for multi-turn RLHF .....

ywang96 committed:
Merge branch 'main' into export-D79840114

ywang96 commented:
> a serious bug for multi-turn RLHF .....

Didn't know this PR was still not merged and thanks for bumping it - I'll get it merged soon

Commit:
update

Signed-off-by: Roger Wang <hey@rogerw.io>
Signed-off-by: Roger Wang <hey@rogerw.me>

Commit:
comment

Signed-off-by: Roger Wang <hey@rogerw.me>

Commit:
add details

Signed-off-by: Roger Wang <hey@rogerw.io>
Signed-off-by: Roger Wang <hey@rogerw.me>

ywang96 committed:
Merge branch 'main' into export-D79840114
